crates/sui-bridge-indexer/src/indexer_builder.rs

```diff
@@ -0,0 +1,405 @@
+    where
+        P: Persistent<R> + 'static,
+    {
+        match self.backfill_strategy {
+            BackfillStrategy::Simple => self.storage.register_task(
+                format!("{} - backfill - {}", self.name, TransactionDigest::random()),
+                current_cp,
+                self.start_from_checkpoint as i64,
+            ),
+            BackfillStrategy::Partitioned { task_size } => {
+                while current_cp < self.start_from_checkpoint {
+                    let target_cp = min(current_cp + task_size, self.start_from_checkpoint);
+                    self.storage.register_task(
+                        format!("{} - backfill - {}", self.name, TransactionDigest::random()),
+                        current_cp,
+                        target_cp as i64,
+                    )?;
+                    current_cp = target_cp;
+                }
+                Ok(())
+            }
+            BackfillStrategy::Disabled => Ok(()),
+        }
+    }
+}
+
+pub trait Persistent<T>: IndexerProgressStore + Sync + Send + Clone {
+    fn write(&self, data: Vec<T>) -> Result<(), Error>;
+}
+
+#[async_trait]
+pub trait IndexerProgressStore: Send {
+    async fn load_progress(&self, task_name: String) -> anyhow::Result<u64>;
+    async fn save_progress(
+        &mut self,
+        task_name: String,
+        checkpoint_number: u64,
+    ) -> anyhow::Result<()>;
+
+    fn tasks(&self, task_prefix: &str) -> Result<Vec<Task>, anyhow::Error>;
+
+    fn register_task(
+        &mut self,
+        task_name: String,
+        checkpoint: u64,
+        target_checkpoint: i64,
+    ) -> Result<(), anyhow::Error>;
+
+    fn update_task(&mut self, task: Task) -> Result<(), anyhow::Error>;
+}
+
+#[async_trait]
+pub trait Datasource<T: Send, P, R> {
+    async fn start_ingestion_task<M>(
+        &self,
+        task_name: String,
+        starting_checkpoint: u64,
+        target_checkpoint: u64,
+        mut storage: P,
+        data_mapper: M,
+    ) -> Result<(), Error>
+    where
+        M: DataMapper<T, R> + 'static,
+        P: Persistent<R> + 'static,
+    {
+        let (join_handle, mut data_channel) = self
+            .create_data_channel(task_name.clone(), starting_checkpoint, target_checkpoint)
+            .await?;
+        while let Some(data) = data_channel.recv().await {
+            if !data.is_empty() {
+                // Ok to unwrap here, checked if the data is empty above.
+                let block_number = Self::get_block_number(data.first().unwrap());
+                let processed_data = data.into_iter().try_fold(vec![], |mut result, d| {
+                    result.append(&mut data_mapper.map(d)?);
+                    Ok::<Vec<_>, Error>(result)
+                })?;
+                storage.write(processed_data)?;
+                storage
+                    .save_progress(task_name.clone(), block_number)
+                    .await?;
```

Comment:
can we add a TODO here to explain the opportunity to do these two writes in one DB transaction, or, update save_progress less frequently? (for better backfilling performance)

in particular, it's not clear to me whether we should separate `IndexerProgressStore` and `Persistent` traits